### ISSUE TITLE
Backup Creation broken on fresh install in minikube

### ISSUE THREAD

**Issue by Toasterson** [CrunchyData/postgres-operator#1234]
On a fresh install in minikube all Backup types fail. pgrestbackup cannot find the WAL archive.

**To Reproduce**
1. `pgo create cluster -n pgodevtest hippo`
1. `pgo backup hippo -n pgodevtest --backup-opts="--type=full --archive-timeout=180"`

**Expected behavior**
Backup suceeds

**Screenshots**
```text
time="2020-01-20T16:39:45Z" level=info msg="pgo-backrest starts"
time="2020-01-20T16:39:45Z" level=info msg="debug flag set to false"
time="2020-01-20T16:39:45Z" level=info msg="backrest backup command requested"
time="2020-01-20T16:39:45Z" level=info msg="command to execute is [pgbackrest backup --type=full --archive-timeout=300 --db-host=10.1.0.10 --db-path=/pgdata/hippo]"
time="2020-01-20T16:39:45Z" level=info msg="command is pgbackrest backup --type=full --archive-timeout=300 --db-host=10.1.0.10 --db-path=/pgdata/hippo "
time="2020-01-20T16:44:55Z" level=error msg="command terminated with exit code 82"
time="2020-01-20T16:44:55Z" level=info msg="output=[]"
time="2020-01-20T16:44:55Z" level=info msg="stderr=[WARN: option repo1-retention-full is not set, the repository may run out of space\n      HINT: to retain full backups indefinitely (without warning), set option 'repo1-retention-full' to the maximum.\nWARN: aborted backup 20200120-152920F of same type exists, will be cleaned to remove invalid files and resumed\nERROR [082]: : could not find WAL segment 00000001000000000000000F after 300 second(s)\nHINT: is archive_command configured correctly?\nHINT: use the check command to verify that PostgreSQL is archiving.\n]"
time="2020-01-20T16:44:55Z" level=error msg="command terminated with exit code 82"
```

**Please tell us about your environment:**

* Operating System: minikube
* Where is this running ( Local, Cloud Provider) minikube (Virtualbox)
* Storage being used (NFS, Hostpath, Gluster, etc): hostpath
* Container Image Tag: 4.2.0
* PostgreSQL Version: 12.1
* Platform (Docker, Kubernetes, OpenShift): Kubernetes CRIO minikube
* Platform Version: 1.17.1

**Additional context**

**Comment by jkatz** [comment]
This is hitting an [archive timeout](https://github.com/pgbackrest/pgbackrest/search?q=ArchiveTimeoutError&unscoped_q=ArchiveTimeoutError) error.

I would need to better understand your environment and how things are deployed.

There is a telltale error message here:

```
could not find WAL segment 00000001000000000000000F after 300 second(s)\nHINT: is archive_command configured correctly?
HINT: use the check command to verify that PostgreSQL is archiving.
```

which we do enable this by default in the containers under management by the Operator. A few questions to start

1. How soon after you can the `pgo create cluster` command did you run the backup?
1. Note, by default we take a full backup after the cluster is available and healthy, as this is an essential part of the high-availability system

**Comment by Toasterson** [comment]
Roughly 5 minutes after.

**Comment by jkatz** [comment]
Can you please provide a `kubectl describe` on your PostgreSQL primary instance pod as well as your pgbackrest repository pods?

From your environment I would guess said commands would resemble:

```shell
kubectl -n pgodevtest describe pods -l pg-cluster=hippo,role=master
kubectl -n pgodevtest describe pods -l pgo-backrest-repo=true,name=hippo-backrest-shared-repo
```

**Comment by Toasterson** [comment]
Gladly:

```text
Name:         hippo-7655689469-x5w8j
Namespace:    pgodevtest
Priority:     0
Node:         minikube/192.168.99.102
Start Time:   Mon, 20 Jan 2020 16:27:48 +0100
Labels:       archive-timeout=60
              crunchy-pgha-scope=hippo
              crunchy_collect=false
              deployment-name=hippo
              name=hippo
              pg-cluster=hippo
              pg-cluster-id=cc665e0a-3696-4aa2-aee8-7ed56ef554a0
              pg-pod-anti-affinity=
              pgo-pg-database=true
              pgo-version=4.2.0
              pgouser=admin
              pod-template-hash=7655689469
              role=master
              service-name=hippo
              vendor=crunchydata
              workflowid=980ee842-edf0-4b3a-a232-991beafd69d5
Annotations:  status:
                {"conn_url":"postgres://10.1.0.10:5432/postgres","api_url":"http://10.1.0.10:8009/patroni","state":"running","role":"master","version":"1....
Status:       Running
IP:           10.1.0.10
IPs:
  IP:           10.1.0.10
Controlled By:  ReplicaSet/hippo-7655689469
Containers:
  database:
    Container ID:   cri-o://0292d656cf77ea7e417a87f1cb50700f2b825e5857893dc30350781bcc712e2e
    Image:          crunchydata/crunchy-postgres-ha:centos7-12.1-4.2.0
    Image ID:       docker.io/crunchydata/crunchy-postgres-ha@sha256:370dc75b882bccfbd24ed814c48b62cee91aa35428899d55c8e5deb1ca89f234
    Ports:          5432/TCP, 8009/TCP
    Host Ports:     0/TCP, 0/TCP
    State:          Running
      Started:      Mon, 20 Jan 2020 16:28:29 +0100
    Ready:          True
    Restart Count:  0
    Liveness:       exec [/opt/cpm/bin/pgha-liveness.sh] delay=30s timeout=10s period=15s #success=1 #failure=3
    Readiness:      exec [/opt/cpm/bin/pgha-readiness.sh] delay=15s timeout=1s period=10s #success=1 #failure=3
    Environment:
      PGHA_PG_PORT:                       5432
      PGHA_USER:                          postgres
      PGHA_INIT:                          <set to the key 'init' of config map 'hippo-pgha-default-config'>  Optional: false
      PATRONI_POSTGRESQL_DATA_DIR:        /pgdata/hippo
      PGBACKREST_STANZA:                  db
      PGBACKREST_REPO1_HOST:              hippo-backrest-shared-repo
      BACKREST_SKIP_CREATE_STANZA:        true
      PGHA_PGBACKREST:                    true
      PGBACKREST_REPO1_PATH:              /backrestrepo/hippo-backrest-shared-repo
      PGBACKREST_DB_PATH:                 /pgdata/hippo
      ENABLE_SSHD:                        true
      PGBACKREST_LOG_PATH:                /tmp
      PGBACKREST_PG1_SOCKET_PATH:         /tmp
      PGBACKREST_PG1_PORT:                5432
      PGBACKREST_REPO_TYPE:               posix
      PGHA_PGBACKREST_LOCAL_S3_STORAGE:   false
      PGHA_DATABASE:                      userdb
      PGHA_CRUNCHYADM:                    true
      PGHA_REPLICA_REINIT_ON_START_FAIL:  true
      PGHA_SYNC_REPLICATION:              false
      PATRONI_KUBERNETES_NAMESPACE:       pgodevtest (v1:metadata.namespace)
      PATRONI_KUBERNETES_SCOPE_LABEL:     crunchy-pgha-scope
      PATRONI_SCOPE:                       (v1:metadata.labels['crunchy-pgha-scope'])
      PATRONI_KUBERNETES_LABELS:          {vendor: "crunchydata"}
      PATRONI_LOG_LEVEL:                  INFO
      PGHOST:                             /tmp
    Mounts:
      /backrestrepo from backrestrepo (rw)
      /crunchyadm from crunchyadm (rw)
      /pgconf from pgconf-volume (rw)
      /pgconf/pgreplicator from primary-volume (rw)
      /pgconf/pgsuper from root-volume (rw)
      /pgconf/pguser from user-volume (rw)
      /pgdata from pgdata (rw)
      /pgwal from pgwal-volume (rw)
      /recover from recover-volume (rw)
      /sshd from sshd (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from pgo-pg-token-xgcwn (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  pgdata:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  hippo
    ReadOnly:   false
  user-volume:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hippo-testuser-secret
    Optional:    false
  primary-volume:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hippo-primaryuser-secret
    Optional:    false
  collect-volume:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  sshd:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hippo-backrest-repo-config
    Optional:    false
  root-volume:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hippo-postgres-secret
    Optional:    false
  pgwal-volume:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  recover-volume:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  report:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  backrestrepo:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  crunchyadm:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  pgconf-volume:
    Type:               Projected (a volume that contains injected data from multiple sources)
    ConfigMapName:      hippo-pgha-default-config
    ConfigMapOptional:  0xc0004717f9
  pgo-pg-token-xgcwn:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  pgo-pg-token-xgcwn
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

```text
Name:         hippo-backrest-shared-repo-5c747dd745-dr566
Namespace:    pgodevtest
Priority:     0
Node:         minikube/192.168.99.102
Start Time:   Mon, 20 Jan 2020 16:27:48 +0100
Labels:       name=hippo-backrest-shared-repo
              pg-cluster=hippo
              pgo-backrest-repo=true
              pod-template-hash=5c747dd745
              service-name=hippo-backrest-shared-repo
              vendor=crunchydata
Annotations:  <none>
Status:       Running
IP:           10.1.0.9
IPs:
  IP:           10.1.0.9
Controlled By:  ReplicaSet/hippo-backrest-shared-repo-5c747dd745
Containers:
  database:
    Container ID:   cri-o://caa5811a323706def59417a689acf80611980b1db78f77b3d81b1aa10f2f0bea
    Image:          crunchydata/pgo-backrest-repo:centos7-4.2.0
    Image ID:       docker.io/crunchydata/pgo-backrest-repo@sha256:5db0f858c631abc60596a1eda7cda444c67181aa14aef4fe43a38f6759883de5
    Port:           2022/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Mon, 20 Jan 2020 16:28:00 +0100
    Ready:          True
    Restart Count:  0
    Environment:
      PGBACKREST_STANZA:           db
      SSHD_PORT:                   2022
      PGBACKREST_DB_PATH:          /pgdata/hippo
      PGBACKREST_REPO_PATH:        /backrestrepo/hippo-backrest-shared-repo
      PGBACKREST_PG1_PORT:         5432
      PGBACKREST_LOG_PATH:         /tmp
      PGBACKREST_PG1_SOCKET_PATH:  /tmp
      PGBACKREST_DB_HOST:          hippo
    Mounts:
      /backrestrepo from backrestrepo (rw)
      /sshd from sshd (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  sshd:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  hippo-backrest-repo-config
    Optional:    false
  backrestrepo:
    Type:        PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:   hippo-pgbr-repo
    ReadOnly:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>

```

**Comment by jkatz** [comment]
So I will now give the preface that we don't typically test in minikube or with hostpath as we try to handle most of our QA towards production environments. That said, I am curious to figure out what's going on :wink: Continuing:

What happens if you try to run `pgbackrest` from the primary pod?

```shell
kubectl exec -n pgodevtest hippo-7655689469-x5w8j pgbackrest info
```

Does this return successfully?

**Comment by Toasterson** [comment]
No I guess minikube does something different with DNS than you are expecting somehow? Although coreDNS should provide a DNS for the nodes and is configured.

```text
ERROR: [125]: remote-0 process on 'hippo-backrest-shared-repo' terminated unexpectedly [255]: ssh: Could not resolve hostname hippo-backrest-shared-repo: Name or service not known
command terminated with exit code 125
```

**Comment by jkatz** [comment]
Ah ha, based on the above I thought it might be the DNS, so this confirms it :smile: 

We're not doing anything fancy, just what Kube expects when using a DNS layer. We're using the service name as part of the pgBackRest request (which is being handled over ssh). I'd suggest checking if the service names are respected as part of the DNS rules in your environment.

I can try to set up minikube in my environment and play with it later and see if I can reproduce it.

**Comment by jkatz** [comment]
@Toasterson Alright, so I spun up a Minikube locally (I used the one that comes with the MacOS Docker package). I followed the quickstart instructions and used `hostpathstorage`. (I know you used the CRI-O runtime for your example).

The initial backup completes as does the subsequent backup, and the WAL archives appear to be pushing correctly. I was able to confirm connectivity between my PostgreSQL and pgBackRest repo Pods with the command I provided in [this comment](https://github.com/CrunchyData/postgres-operator/issues/1234#issuecomment-576387486).

My suggestion would be to check your DNS to ensure everything is set up and running OK. Outside of that, I don't believe there is an issue with the Operator codebase.

If you're continuing to have trouble and believe you've hit upon a bug, please let us know. Thanks!

**Comment by Toasterson** [comment]
I found the problem in one of the minikube command line switches I used to deploy my minkube instance `--dns-domain` This sets the kubelets cluster domain to the specified Value. Either that does not work as expected with postgres operator or the Feature is broken in minikube. Leaving it out gets me a working Installation. I am leaving this here as an FYI in case production installs are also able to set the cluster domain somehow. I found no indication kubeadm or others are capable of doing so in a quick search.

I must say with the working installation I really like postgres operator and will be using it for production later. Thanks for the in depth Support and keep up the good work.
